File: flutter/lib/camera/face_analyzer.dart
import 'dart:developer';
import 'dart:typed_data';

import 'package:camera/camera.dart';
import 'package:flutter/material.dart';
import 'package:google_mlkit_face_detection/google_mlkit_face_detection.dart';
import 'package:liveness/camera/camera_view.dart';
import 'package:liveness/camera/nv21.dart';

enum LivenessStep { front, smile, side, done }

enum FaceError { notCenter, tooFar, tooClose, multipleFaces, none }

class _LivenessController {
  var _currentStep = LivenessStep.front;
  var _retryCount = 0;
  final _maxRetries = 5;

  LivenessStep get step => _currentStep;

  void reset() {
    _currentStep = LivenessStep.front;
    _retryCount = 0;
  }

  void nextStep() {
    _retryCount = 0;
    _currentStep = switch (_currentStep) {
      LivenessStep.front => LivenessStep.smile,
      LivenessStep.smile => LivenessStep.side,
      LivenessStep.side => LivenessStep.done,
      LivenessStep.done => LivenessStep.done,
    };
  }

  void onFailedDetection() {
    _retryCount++;
    if (_retryCount > _maxRetries) {
      reset();
    }
  }
}

class FaceAnalyzer {
  final Function(LivenessStep step, FaceError error) statusCallback;
  final Function(LivenessStep step, XFile file) onCapture;
  final VoidCallback onDone;

  final _controller = _LivenessController();
  final GlobalKey<CameraViewState> cameraKey;

  final _dector = FaceDetector(
    options: FaceDetectorOptions(
      enableContours: false,
      enableLandmarks: false,
      enableClassification: true,
      minFaceSize: 0.15,
      enableTracking: false,
      performanceMode: FaceDetectorMode.fast,
    ),
  );

  FaceAnalyzer({
    required this.cameraKey,
    required this.statusCallback,
    required this.onCapture,
    required this.onDone,
  });

  var _stepSuccessTime = 0;

  void handleFailure(LivenessStep step, FaceError error) {
    _stepSuccessTime = 0;
    statusCallback(step, error);
    _controller.onFailedDetection();
  }

  Future<void> analyze(CameraImage cameraImage) async {
    final step = _controller.step;
    if (step == LivenessStep.done) return;
    final inputImage = _inputImageFromCameraImage(cameraImage);
    if (inputImage == null) {
      handleFailure(step, FaceError.none);
      return;
    }

    final faces = await _dector.processImage(inputImage);
    if (faces.isEmpty) {
      handleFailure(step, FaceError.none);
      return;
    } else if (faces.length > 1) {
      handleFailure(step, FaceError.multipleFaces);
      return;
    }

    final rotation = inputImage.metadata.rotation;
    final imageSize = inputImage.metadata.size;
    final reverseWH =
        rotation == InputImageRotation.rotation90deg ||
        rotation == InputImageRotation.rotation270deg;
    final frameW = reverseWH ? imageSize.height : imageSize.width;
    final frameH = reverseWH ? imageSize.width : imageSize.height;

    final face = faces.first;
    final faceRect = clampRect(face.boundingBox, frameH, frameH);

    // 面部位置 & 距离检测
    if (step == LivenessStep.front) {
      final distanceError = _detectFaceDistance(faceRect, frameW, frameH);
      if (distanceError != FaceError.none) {
        handleFailure(step, distanceError);
        return;
      }

      final positionError = _detectFacePosition(faceRect, frameW, frameH);
      if (positionError != FaceError.none) {
        handleFailure(step, positionError);
        return;
      }
    }

    statusCallback(step, FaceError.none);

    final yaw = face.headEulerAngleY ?? 0.0;
    final pitch = face.headEulerAngleZ ?? 0.0;

    final success = switch (step) {
      LivenessStep.front =>
        yaw >= -12.0 && yaw <= 12.0 && pitch >= -8.0 && pitch <= 8.0,
      LivenessStep.smile => (face.smilingProbability ?? 0.0) > 0.3,
      LivenessStep.side => yaw < -20.0 || yaw > 20.0,
      LivenessStep.done => false,
    };

    if (success) {
      final now = DateTime.now().millisecondsSinceEpoch;
      if (_stepSuccessTime == 0) {
        _stepSuccessTime = now;
      } else {
        final elapsedTime = now - _stepSuccessTime;
        final delayTime = switch (step) {
          LivenessStep.front => 800,
          LivenessStep.smile => 300,
          LivenessStep.side => 100,
          LivenessStep.done => 0,
        };
        if (elapsedTime >= delayTime) {
          _stepSuccessTime = 0;
          final file = await _capturePhoto();
          if (file == null) return;
          onCapture(step, file);
          _controller.nextStep();
          if (_controller.step == LivenessStep.done) {
            onDone();
          }
        }
      }
    } else {
      _stepSuccessTime = 0;
    }
  }

  Future<XFile?> _capturePhoto() async {
    try {
      return cameraKey.currentState?.takePicture();
    } catch (e) {
      log('Error capturing photo', error: e);
    }
    return null;
  }

  void dispose() {
    _dector.close();
  }

  FaceError _detectFaceDistance(Rect faceRect, double frameW, double frameH) {
    final faceRatio = (faceRect.width * faceRect.height) / (frameW * frameH);
    final tooCloseRatio = 0.36;
    final tooFarRatio = 0.12;
    if (faceRatio > tooCloseRatio) {
      return FaceError.tooClose;
    } else if (faceRatio < tooFarRatio) {
      return FaceError.tooFar;
    }
    return FaceError.none;
  }

  FaceError _detectFacePosition(Rect faceRect, double frameW, double frameH) {
    final centerTolerance = 0.15;
    final center = faceRect.center;
    final dxRatio = (center.dx - frameW / 2) / frameW;
    final dyRatio = (center.dy - frameH / 2) / frameH;
    if (dxRatio.abs() > centerTolerance || dyRatio.abs() > centerTolerance) {
      return FaceError.notCenter;
    }
    return FaceError.none;
  }

  Rect clampRect(Rect r, double frameW, double frameH) {
    final l = r.left.clamp(0, frameW).toDouble();
    final t = r.top.clamp(0, frameH).toDouble();
    final rr = r.right.clamp(0, frameW).toDouble();
    final bb = r.bottom.clamp(0, frameH).toDouble();
    return Rect.fromLTRB(l, t, rr, bb);
  }

  InputImage? _inputImageFromCameraImage(CameraImage image) {
    // get image rotation
    // it is used in android to convert the InputImage from Dart to Java: https://github.com/flutter-ml/google_ml_kit_flutter/blob/master/packages/google_mlkit_commons/android/src/main/java/com/google_mlkit_commons/InputImageConverter.java
    // `rotation` is not used in iOS to convert the InputImage from Dart to Obj-C: https://github.com/flutter-ml/google_ml_kit_flutter/blob/master/packages/google_mlkit_commons/ios/Classes/MLKVisionImage%2BFlutterPlugin.m
    // in both platforms `rotation` and `camera.lensDirection` can be used to compensate `x` and `y` coordinates on a canvas: https://github.com/flutter-ml/google_ml_kit_flutter/blob/master/packages/example/lib/vision_detector_views/painters/coordinates_translator.dart
    final sensorOrientation = cameraKey.currentState?.sensorOrientation;
    if (sensorOrientation == null) return null;
    InputImageRotation? rotation = InputImageRotationValue.fromRawValue(
      sensorOrientation,
    );
    if (rotation == null) return null;

    final format = InputImageFormatValue.fromRawValue(image.format.raw);
    if (format == null) return null;

    Uint8List? bytes;
    if (format == InputImageFormat.yuv420 ||
        format == InputImageFormat.yuv_420_888) {
      if (image.planes.length != 3) return null;
      bytes = image.getNv21Uint8List();
    } else if (format == InputImageFormat.nv21 ||
        format == InputImageFormat.bgra8888) {
      bytes = image.planes.first.bytes;
    }

    if (bytes == null) return null;
    // compose InputImage using bytes
    return InputImage.fromBytes(
      bytes: bytes,
      metadata: InputImageMetadata(
        size: Size(image.width.toDouble(), image.height.toDouble()),
        rotation: rotation, // used only in Android
        format: format, // used only in iOS
        bytesPerRow: image.planes.first.bytesPerRow, // used only in iOS
      ),
    );
  }
}

File: rn/src/FaceAnalyzer.ts
import { Camera, PhotoFile } from 'react-native-vision-camera';
import {
  Bounds,
  Face,
  FaceDetectionOptions,
} from 'react-native-vision-camera-face-detector';

export type LivenessStep = 'front' | 'smile' | 'side' | 'done';
export type FaceError =
  | 'not_center'
  | 'too_far'
  | 'too_close'
  | 'multiple_faces'
  | 'none';

class LivenessController {
  private currentStep: LivenessStep = 'front';
  private retryCount: number = 0;
  private maxRetries: number = 5;

  private reset() {
    this.currentStep = 'front';
    this.retryCount = 0;
  }

  nextStep() {
    this.retryCount = 0;
    if (this.currentStep === 'front') {
      this.currentStep = 'smile';
    } else if (this.currentStep === 'smile') {
      this.currentStep = 'side';
    } else if (this.currentStep === 'side') {
      this.currentStep = 'done';
    }
  }

  public onFailedDetection() {
    this.retryCount++;
    if (this.retryCount > this.maxRetries) {
      this.reset();
    }
  }

  public getStep(): LivenessStep {
    return this.currentStep;
  }
}

export class FaceAnalyzer {
  private camera: React.RefObject<Camera | null>;
  private statusCallback: (step: LivenessStep, error: FaceError) => void;
  private onDone: (files: PhotoFile[]) => void;

  constructor(
    camera: React.RefObject<Camera | null>,
    statusCallback: (step: LivenessStep, error: FaceError) => void,
    onDone: (files: PhotoFile[]) => void,
  ) {
    this.camera = camera;
    this.statusCallback = statusCallback;
    this.onDone = onDone;
  }

  public faceDetectionOptions: FaceDetectionOptions = {
    cameraFacing: 'front',
    classificationMode: 'all',
    performanceMode: 'fast',
    contourMode: 'none',
    landmarkMode: 'none',
    minFaceSize: 0.15,
    trackingEnabled: false,
  };
  private controller = new LivenessController();
  private stepSuccessTime = 0;
  private isBusy = false;

  private stepDuration: Record<LivenessStep, number> = {
    front: 500,
    smile: 100,
    side: 100,
    done: 0,
  };

  private stepFiles: Partial<Record<LivenessStep, PhotoFile>> = {};

  private handleFailure(step: LivenessStep, error: FaceError) {
    this.stepSuccessTime = 0;
    this.statusCallback(step, error);
    this.controller.onFailedDetection();
  }

  public async analyze(faces: Face[], frameW: number, frameH: number) {
    const step = this.controller.getStep();
    console.log('step', step, 'isBusy', this.isBusy);
    if (this.isBusy || step === 'done') return;
    this.isBusy = true;

    try {
      if (faces.length === 0) {
        this.handleFailure(step, 'none');
        return;
      } else if (faces.length > 1) {
        this.handleFailure(step, 'multiple_faces');
        return;
      }

      const face = faces[0];
      const faceBounds = this.clampBounds(face, frameW, frameH);

      // 面部位置 & 距离检测
      if (step === 'front') {
        const distanceError = this.detectFaceDistance(
          faceBounds,
          frameW,
          frameH,
        );
        if (distanceError) {
          this.handleFailure(step, distanceError);
          return;
        }
        const positionError = this.detectFacePosition(
          faceBounds,
          frameW,
          frameH,
        );
        if (positionError) {
          this.handleFailure(step, positionError);
          return;
        }
      }

      this.statusCallback(step, 'none');

      const yaw = face.yawAngle;
      const pitch = face.pitchAngle;
      let success = false;
      if (step === 'front') {
        success = yaw >= -12 && yaw <= 12 && pitch >= -8 && pitch <= 8;
      } else if (step === 'smile') {
        success = face.smilingProbability > 0.3;
      } else if (step === 'side') {
        success = yaw < -20.0 || yaw > 20.0;
      }
      console.log(step, 'success', success);

      if (success) {
        const now = Date.now();
        if (this.stepSuccessTime === 0) {
          this.stepSuccessTime = now;
        } else {
          const elapsedTime = now - this.stepSuccessTime;
          if (elapsedTime >= this.stepDuration[step]) {
            this.stepSuccessTime = 0;
            const file = await this.camera.current?.takeSnapshot();
            if (file) {
              this.stepFiles[step] = file;
              this.controller.nextStep();
              if (this.controller.getStep() === 'done') {
                const files = [
                  this.stepFiles.front,
                  this.stepFiles.smile,
                  this.stepFiles.side,
                ] as PhotoFile[];
                this.onDone(files);
              }
            }
          }
        }
      } else {
        this.stepSuccessTime = 0;
      }
    } catch (e) {
      console.log('error', e);
    } finally {
      this.isBusy = false;
    }
  }

  private detectFaceDistance(
    faceBounds: Bounds,
    frameW: number,
    frameH: number,
  ): FaceError | null {
    const faceRatio =
      (faceBounds.width * faceBounds.height) / (frameW * frameH);
    const tooCloseRatio = 0.36;
    const tooFarRatio = 0.12;
    if (faceRatio > tooCloseRatio) {
      return 'too_close';
    } else if (faceRatio < tooFarRatio) {
      return 'too_far';
    }
    return null;
  }

  private detectFacePosition(
    faceBounds: Bounds,
    frameW: number,
    frameH: number,
  ): FaceError | null {
    const centerTolerance = 0.15;
    const centerX = faceBounds.x + faceBounds.width / 2;
    const centerY = faceBounds.y + faceBounds.height / 2;
    const dxRatio = (centerX - frameW / 2) / frameW;
    const dyRatio = (centerY - frameH / 2) / frameH;
    if (
      Math.abs(dxRatio) > centerTolerance ||
      Math.abs(dyRatio) > centerTolerance
    ) {
      return 'not_center';
    }
    return null;
  }

  private clampBounds(face: Face, frameW: number, frameH: number) {
    const bounds = face.bounds;
    const x = Math.max(0, bounds.x);
    const y = Math.max(0, bounds.y);
    const w = Math.min(bounds.width, frameW - x);
    const h = Math.min(bounds.height, frameH - y);
    return { x, y, width: w, height: h };
  }
}
